### PR TITLE
Add docstrings

### PR DESCRIPTION
**Pull request by frnmst**
- Add more docstrings in `narwhals.dataframe.DataFrame` class
- Fix indentation in pre-commit file

### CODE REVIEW

File: narwhals/dataframe.py [hunk 1]
```diff
@@ -477,18 +485,353 @@ def rename(self, mapping: dict[str, str]) -> Self:
...
 
     def head(self, n: int) -> Self:
+        r"""
+        Get the first `n` rows.
+
+        Arguments:
+            n: Number of rows to return. If a negative value is passed, return all rows
+                except the last `abs(n)`.
+
+        See Also: `tail`, `glimpse`, `slice`
```
Comment (MarcoGorelli) [inline narwhals/dataframe.py:495]: I don't this we have these yet, shall we remove this?

File: narwhals/dataframe.py [hunk 2]
```diff
@@ -477,18 +485,353 @@ def rename(self, mapping: dict[str, str]) -> Self:
...
+        """
         return super().head(n)
 
     def drop(self, *columns: str | Iterable[str]) -> Self:
+        r"""
+        Remove columns from the dataframe.
+
+        Arguments:
+            *columns: Names of the columns that should be removed from the dataframe.
+                       Accepts column selector input.
```
Comment (MarcoGorelli) [inline narwhals/dataframe.py:560]: We don't yet have column selectors in Narwhals, let's remove this for now

File: narwhals/dataframe.py [hunk 3]
```diff
@@ -477,18 +485,353 @@ def rename(self, mapping: dict[str, str]) -> Self:
...
+            shape: (3, 1)
+            ┌─────┐
+            │ ham │
+            │ --- │
+            │ str │
+            ╞═════╡
+            │ a   │
+            │ b   │
+            │ c   │
+            └─────┘
```
Comment (MarcoGorelli) [inline narwhals/dataframe.py:637]: let's remove this one

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
These docstrings are great, thank you so much @frnmst !

Looks like CI is failing with
```
E   ValueError: line 29 of the doctest for narwhals.dataframe.DataFrame.group_by has an invalid option: '+IGNORE_RESULT'
```

Just got some minor comments

**Comment by frnmst** [comment]
I think I fixed everything

**Review by MarcoGorelli — APPROVED** [review]
Thank you @frnmst ! Seriously good docstrings here, well done!

Little tip - I'd suggest making a new branch for each new PR (rather than opening PRs from `main`), it just makes it a bit easier to review

**Comment by MarcoGorelli** [comment]
Also, I removed some `doctest skip`s, as the doctests pass without them. For group-by, we can get a consistent output by adding a `sort`

**Comment by frnmst** [comment]
> Thank you @frnmst ! Seriously good docstrings here, well done!
> 
> Little tip - I'd suggest making a new branch for each new PR (rather than opening PRs from `main`), it just makes it a bit easier to review


Ok, will do


> Also, I removed some `doctest skip`s, as the doctests pass without them. For group-by, we can get a consistent output by adding a `sort`

Ok, I'll keep that in mind for the next docstrings

**Pull request merged**